### ISSUE TITLE
Add data "streaming"/"buffering"/"chunking" to Neat

### ISSUE THREAD

**Issue by luiscarbonell** [liquidcarrot/carrot#85]
**Is your feature request related to a problem? Please describe.**

No.

**Describe the solution you'd like**

Being able to "stream" data into a population as if it were an individual network looping through `network.activate()` and `network.propagate()`.

Alternatively we can begin to support RL

**Additional context**

![image](https://user-images.githubusercontent.com/21148993/58844940-4b8f6700-8647-11e9-81ff-9dd7c1165f66.png)

**Comment by tracycollins** [comment]
Would love to be able to stream a data set into the evolve() function. I'm running evolution on multiple machines and each one needs to load in the entire data set before evolution can start (and i generate the data set on one machine and send it to the others in the cluster). As my data set continues to grow, this will not be possible at some point. 

How are others handling large data sets now?

**Comment by christianechevarria** [comment]
@tracycollins thanks for sharing with us! We'll make this a much higher priority as it seems like something a lot of people are looking for

**Comment by tracycollins** [comment]
> @tracycollins thanks for sharing with us! We'll make this a much higher priority as it seems like something a lot of people are looking for

Looking forward to this.

Also, an unrelated item that would be nice to have: labels for input and output nodes. I think they are only identified by array index right now? It would help with testing various networks in parallel that have different inputs.

**Comment by luiscarbonell** [comment]
@tracycollins Out of curiosity what are you using to add the additional "object wrapping" functionality? 

_I was thinking of adding this using [`mappr`](https://www.npmjs.com/package/mappr) or similar functionality._

**Comment by tracycollins** [comment]
@luiscarbonell I just used plain ol javascript to wrap the network object to add meta data as well as input node names.

I didn't know about mappr until you mentioned it. I'll give it a look.

**Comment by luiscarbonell** [comment]
@tracycollins I went ahead and created a [project](https://github.com/liquidcarrot/carrot/projects/18?add_cards_query=is%3Aopen) so I can start "knocking it out": https://github.com/liquidcarrot/carrot/projects/18

Do you have an ideal "interface" you would like?

**Example**
>Maybe something like this....?

```javascript
const { Neat } = require("@liquid-carrot/carrot")

const neat = new Neat();

neat.evolveFromPath("./filepath")
  .on("iteration", (stuff) => {
    console.log("Just ran one generation");
  })
  .on("finish", (information) => {
    console.log("Just finished training");
  })
```

Let me know what you think/want to see - is there any you prefer?

**Comment by tracycollins** [comment]
@luiscarbonell this looks good, although I'd also like an option to feed it via a node stream, and maybe also a way to feed it via a fifo, in addition to reading from a file path.  That would allow me to generate data on the fly

**Comment by tracycollins** [comment]
@luiscarbonell And, if fed by a stream or fifo, there needs to be a way to signal the end of the data set.

**Comment by luiscarbonell** [comment]
@tracycollins 

```javascript
```

**Comment by tracycollins** [comment]
@luiscarbonell Yes!

**Comment by tracycollins** [comment]
@luiscarbonell How would training end using node stream? close the stream?

**Comment by luiscarbonell** [comment]
>@luiscarbonell How would training end using node stream? close the stream?

@tracycollins Yeah, if I remember correctly, Node ends streams by passing `null` - I think the WebStream MDN API uses the same.

I'll document it more accurately when we're done implementing it.
